File: sdk/typescript/examples/32-human-guardrail.ts
/**
 * 32 - Human-in-the-loop Guardrail (onFail='human')
 *
 * Demonstrates a guardrail that pauses the workflow for human review when
 * the output fails validation. The human can approve, reject, or edit.
 *
 * Since the workflow pauses at a HumanTask, this example uses start()
 * (async) instead of run() (blocking).
 *
 * Requirements:
 *   - Conductor server with LLM support
 *   - AGENTSPAN_SERVER_URL=http://localhost:8080/api as environment variable
 *   - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
 */

import { z } from 'zod';
import { Agent, AgentRuntime, guardrail, tool } from '../src/index.js';
import type { GuardrailResult } from '../src/index.js';
import { llmModel } from './settings.js';

// -- Guardrail ---------------------------------------------------------------

const complianceCheck = guardrail(
  (content: string): GuardrailResult => {
    const flaggedTerms = ['investment advice', 'guaranteed returns', 'risk-free'];
    for (const term of flaggedTerms) {
      if (content.toLowerCase().includes(term.toLowerCase())) {
        return {
          passed: false,
          message: `Response contains flagged term: '${term}'. Needs human review.`,
        };
      }
    }
    return { passed: true };
  },
  {
    name: 'compliance',
    position: 'output',
    onFail: 'human',
  },
);

// -- Tool --------------------------------------------------------------------

const getMarketData = tool(
  async (args: { ticker: string }) => {
    return {
      ticker: args.ticker,
      price: 185.42,
      change: '+2.3%',
      volume: '45.2M',
    };
  },
  {
    name: 'get_market_data',
    description: 'Get current market data for a stock ticker.',
    inputSchema: z.object({
      ticker: z.string().describe('The stock ticker symbol'),
    }),
  },
);

// -- Agent -------------------------------------------------------------------

export const agent = new Agent({
  name: 'finance_agent',
  model: llmModel,
  tools: [getMarketData],
  instructions:
    'You are a financial information assistant. Provide market data ' +
    'and general financial information. You may discuss investment ' +
    'strategies and returns.',
  guardrails: [complianceCheck],
});

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('32-human-guardrail.ts') || process.argv[1]?.endsWith('32-human-guardrail.js')) {
  const runtime = new AgentRuntime();
  try {
    // Start the agent (async -- does not block)
    const handle = await runtime.start(
      agent,
      'Look up AAPL and explain whether it\'s a good investment. ' +
      'Include your opinion on potential returns.',
    );
    console.log(`Workflow started: ${handle.workflowId}`);

    // Poll for status
    for (let i = 0; i < 60; i++) {
      const status = await handle.getStatus();
      console.log(`  Status: ${status.status} (waiting=${status.isWaiting})`);

      if (status.isWaiting) {
        console.log('\n--- Workflow paused for human review ---');
        console.log('The guardrail flagged the response for compliance review.');
        console.log('Options: approve(), reject(reason), or respond(output)');

        // In a real app, a human would review in the Conductor UI.
        // Here we auto-reject for the demo.
        console.log('Auto-rejecting for demo...');
        await handle.reject('bad idea');
        console.log('Rejected! Resuming workflow...\n');
      }

      if (status.isComplete) {
        console.log(`\nFinal output: ${status.output}`);
        break;
      }

      await new Promise((resolve) => setTimeout(resolve, 1000));
    }
  } finally {
    await runtime.shutdown();
  }
}

File: sdk/python/examples/32_human_guardrail.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Human-in-the-loop guardrail — ``on_fail="human"``.

Demonstrates a guardrail that pauses the workflow for human review when
the output fails validation.  The human can approve, reject, or edit.

Since the workflow pauses at a HumanTask, this example uses ``start()``
(async) instead of ``run()`` (blocking).

Requirements:
    - Conductor server with LLM support
    - AGENTSPAN_SERVER_URL=http://localhost:6767/api as environment variable
    - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
"""

import time

from agentspan.agents import (
    Agent,
    AgentRuntime,
    Guardrail,
    GuardrailResult,
    OnFail,
    Position,
    guardrail,
    tool,
)
from settings import settings


# ── Guardrail ────────────────────────────────────────────────────────────

@guardrail
def compliance_check(content: str) -> GuardrailResult:
    """Flag any response that mentions specific financial terms for review."""
    flagged_terms = ["investment advice", "guaranteed returns", "risk-free"]
    for term in flagged_terms:
        if term.lower() in content.lower():
            return GuardrailResult(
                passed=False,
                message=f"Response contains flagged term: '{term}'. Needs human review.",
            )
    return GuardrailResult(passed=True)


# ── Tool ─────────────────────────────────────────────────────────────────

@tool
def get_market_data(ticker: str) -> dict:
    """Get current market data for a stock ticker."""
    return {
        "ticker": ticker,
        "price": 185.42,
        "change": "+2.3%",
        "volume": "45.2M",
    }


# ── Agent ────────────────────────────────────────────────────────────────

agent = Agent(
    name="finance_agent",
    model=settings.llm_model,
    tools=[get_market_data],
    instructions=(
        "You are a financial information assistant. Provide market data "
        "and general financial information. You may discuss investment "
        "strategies and returns."
    ),
    guardrails=[
        Guardrail(
            compliance_check,
            position=Position.OUTPUT,
            on_fail=OnFail.HUMAN,
            name="compliance",
        ),
    ],
)


if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(agent, "Look up AAPL and summarize the latest price movement.")
        result.print_result()

        # Production pattern:
        # 1. Deploy once during CI/CD:
        # runtime.deploy(agent)
        # CLI alternative:
        # agentspan deploy --package examples.32_human_guardrail
        #
        # 2. In a separate long-lived worker process:
        # runtime.serve(agent)

        # Interactive human-review alternative:
        # # Start the agent (async — doesn't block)
        # handle = runtime.start(
        #     agent,
        #     "Look up AAPL and explain whether it's a good investment. "
        #     "Include your opinion on potential returns.",
        # )
        # print(f"Workflow started: {handle.execution_id}")

        # # Poll for status
        # for i in range(60):
        #     status = handle.get_status()
        #     print(f"  Status: {status.status} (waiting={status.is_waiting})")

        #     if status.is_waiting:
        #         print("\n--- Workflow paused for human review ---")
        #         print("The guardrail flagged the response for compliance review.")
        #         print("Options: approve(), reject(reason), or respond(output)")

        #         # In a real app, a human would review in the Conductor UI.
        #         # Here we auto-approve for the demo.
        #         print("Auto-approving for demo...")
        #         runtime.reject(handle.execution_id, "bad idea")
        #         print("Approved! Resuming workflow...\n")

        #     if status.is_complete:
        #         print(f"\nFinal output: {status.output}")
        #         break

        #     time.sleep(1)
        # else:
        #     print("Timed out waiting for workflow to complete")
        #     runtime.cancel(handle.execution_id)

